rfcs/0070-target-preprocessing.md

```diff
@@ -0,0 +1,208 @@
+* Architecture Pre-processing - maps `Target` `attrs` to a new `arch` object
+* Keys Pre-processing - maps `Target` `attrs` and `keys` to a new set of `keys`
+
+These new preprocessors will be illustrated using examples targeting TVM for Arm(R) Cortex(R)-M4.
+
+## Architecture Pre-processing
+```c++
+Target("c")
+    .set_arch_preprocessor(MyArchPreprocessor)
+```
+
+This takes the `attrs` from `Target` and converts them into an object representing the architectural features of the `Target`, which can then be accessed using the `GetArch` method similar to `GetAttr`:
+
+```c++
+Target my_target("c -mcpu=cortex-m4");
+my_target->GetArch<Bool>("is_aarch64", false); // false
+my_target->GetArch<Bool>("has_dsp", false); // true
+```
+
+```python
+my_target = Target("c -mcpu=cortex-m4")
+my_target.arch.is_aarch64 // false
+my_target.arch.has_dsp // true
+```
+
+## Keys Pre-processing
+
+```c++
+Target("c")
+    .set_keys_preprocessor(MyKeysPreprocessor)
+```
+
+This takes the `attrs` from `Target` and maps them to relevant `keys` for use when selecting schedules:
+
+```c++
+Target my_target("c -mcpu=cortex-m4");
+my_target->keys; // ["arm_cpu", "cpu"] <-- "cpu" is taken from default keys and merged by the pre-preprocessor
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Currently, there is a single `preprocessor` which takes an input of `attrs` and expects the same `attrs` returned with pre-processing applied:
+
+https://github.com/apache/tvm/blob/d2db9cb0d839e32778f461b77e59f6418282a511/src/target/target.cc#L810-L814
+
+In extension to this, a series of new pre-processors will be defined:
+
+```c++
+using TargetAttrs = Map<String, ObjectRef>;
+using TargetArch = Map<String, ObjectRef>;
+using TargetKeys = Array<String>;
+
+using FTVMAttrPreprocessor = runtime::TypedPackedFunc<TargetAttrs(TargetAttrs)>;
+using FTVMArchPreprocessor = runtime::TypedPackedFunc<TargetArch(TargetAttrs)>;
+using FTVMKeysPreprocessor = runtime::TypedPackedFunc<TargetKeys(TargetAttrs, TargetKeys)>;
+```
+
+These implementations can be stored under `src/target/preprocessors/<arch_identifier>.{cc.h}` to allow them to be composed together such as:
+
+* src/target/preprocessors/aarch64.cc
+* src/target/preprocessors/cpu.cc
+
+Where the `cpu` pre-processor can utilise the `aarch64` pre-processor if detected.
+
+## Rename Attr Preprocessor
+To help avoid confusion between the existing `attrs` `preprocessor` and the new pre-processors, the `attrs` pre-processor will be renamed from `preprocessor` to `attr_preprocessor`:
+
+```c++
+class TargetKind {
+    ...
+    FTVMAttrPreprocessor attr_preprocessor;
+
+    ...
+}
+```
+
+## Architecture Preprocessor
+The first new pre-processor, which processes `attrs` in to an `arch` object, is registered as a new field is added to `TargetKind`:
+
+```c++
+class TargetKind {
+    ...
+    FTVMArchPreprocessor arch_preprocessor;
```

Comment:
Is it's a part of `TargetKind`, then you cannot customize it per specific architecture.  For example, both x86-64 and aarch64 have the same kind: "llvm", but they will need different preprocessors.

Reply:
I believe we'll end up with something similar to the example in `clang` (https://github.com/llvm/llvm-project/blob/2f04e703bff3d9858f53225fa7c780b240c3e247/clang/lib/Driver/ToolChains/Clang.cpp#L331-L343) where the actual parser can take different routes dependent on more specific hints.

Reply:
This is what is meant here:
https://github.com/apache/tvm-rfcs/pull/71/files#diff-c04c196e8069297e1fcfdd5a514a268b971943b265649eaa41edd8d47052d56aR85-R90

Reply:
Clang driver takes flags given by the user and translates them into specific "subtarget" configuration.  There is a flag `-target <triple>`, which roughly corresponds to a `TargetMachine`, and then there are other details (implicit or explicit) that eventually create `TargetSubtargetInfo`.  That part of clang allows each target to specify the exact features, having already established what that target it.  This is similar, for example, to how target kind "rocm" has attributes specific to the ROCm family, the key here being that we know we're dealing with ROCm, and not with any generic GPU.

The target kind "llvm", on the other hand, _does_ correspond to any generic CPU, be it x86 or Arm.  Flags specific to x86 may not be applicable to Arm, and vice versa.  If we attach every arch preprocessor to kind "llvm", then whatever code executes them will have to know which preprocessor to apply, and currently there is no information that would aid with this selection.

This goes beyond the scope of this RFC, but my vote would be to eliminate "llvm" as a target kind, and replace it with specific kinds of CPUs.  We could have target "host" that would correspond to the compilation host (and can be automatically translated to an appropriate specific CPU kind).

Reply:
> Clang driver takes flags given by the user and translates them into specific "subtarget" configuration. There is a flag -target <triple>, which roughly corresponds to a TargetMachine, and then there are other details (implicit or explicit) that eventually create TargetSubtargetInfo.

Apologies, I oversimplified, you're right that Clang/LLVM has a few more data structures to describe this (including `SubTarget`s) - in this RFC we're essentially using `Map<String, ObjectRef>` as a replacement for most of the `SubTarget` in the `.arch` property. I can add further details to the Clang/LLVM Prior Art to reflect this 😸

> The target kind "llvm", on the other hand, does correspond to any generic CPU, be it x86 or Arm. Flags specific to x86 may not be applicable to Arm, and vice versa. If we attach every arch preprocessor to kind "llvm", then whatever code executes them will have to know which preprocessor to apply, and currently there is no information that would aid with this selection.

Sorry, I'm not understanding so I'll provide an example 😸 , in the case of LLVM we'd attach a pre-processor which can inspect the attributes and do something similar to:

```c++
Map<String, ObjectRef> llvm_preprocessor(...attrs...) {
  bool is_aarch64 = ...logic ...;
  if (is_aarch64) {
     return aarch64_preprocessor(...attrs...);
  }
}
```

Therefore, whenever you instantiate a `Target("llvm -march=aarch64-woof-woof-woof")` the preprocessor would give you a `target.arch` `Map` which would be pre-processed using the `aarch64` processor, similar to the snippet I pasted from `clang` with the `switch` logic. This does mean we have to create a preprocessor for both the `llvm` and `c` `Target`s to map `mcpu`, `march`, `mtune` or `mtriple` to the relevant logic path in the same way that LLVM suggests processing `march` and `mcpu` for `SubTargetInfo` - this preprocessor would have the freedom to generate any eventual `Map<String, ObjectRef>` for the specific details of a given architecture which is being targeted by the `llvm` `Target` of TVM.

> This goes beyond the scope of this RFC, but my vote would be to eliminate "llvm" as a target kind, and replace it with specific kinds of CPUs. We could have target "host" that would correspond to the compilation host (and can be automatically translated to an appropriate specific CPU kind).

The `Target` is definitely a difficult problem to unpick, so glad you agree it's out of scope 😸 , personally I would avoid magic `Target`s such as "composite" and "host" in favour of better defining them - `llvm` is maybe just a code generator for however we describe a CPU `Target`? Potentially worth starting an RFC to get the discussion going and then discuss in the community meetup?

Reply:
Looking at your example: `set_arch_preprocessor(MyArchPreprocessor)`, what arch would this apply to?  If I need to add code to the `llvm_preprocessor` above to recognize MyArch, and then call `MyArchPreprocessor`, then there is no point in having the `set_arch_preprocessor` in the first place.

Reply:
My initial thought was that `MyArchPreprocessor` would be the CPU preprocessor which invokes other preprocessors when it decides they're appropriate. So you'd have:

```c++
Map<String, ObjectRef> AArch64PreProcessor(...attrs...) {
  return ...things...;
}

Map<String, ObjectRef> CPUPreprocessor(...attrs...) {
  bool is_aarch64 = ...logic ...;
  if (is_aarch64) {
     return AArch64PreProcessor(...attrs...);
  }
}

TVM_REGISTER_TARGET_KIND("llvm", kDLCPU)
    .set_arch_preprocessor(CPUPreprocessor);
```

Which then means when you invoke:

```python
my_target = Target("llvm -mtriple=aarch64-woof-woof");
```

You can inspect it as:

```python
my_target.arch.is_aarch64 // True
```

This then allows us to add multiple preprocessors to the `llvm` `Target` via the single CPU preprocessor entrypoint.

Reply:
This would be equivalent to adding `arch` as an attribute, and setting it using the attribute preprocessor, wouldn't it?  The difference here is that we don't access `arch` as an attribute, but as a member.

Reply:
> my_target.arch.is_aarch64 // True

I think it would be `my_target.arch["is_aarch64"]`.

Reply:
It'll come out as `my_target.arch.is_aarch64` due to using `__getattr__` here to prettify the attributes:
https://github.com/apache/tvm-rfcs/pull/71/files#diff-c04c196e8069297e1fcfdd5a514a268b971943b265649eaa41edd8d47052d56aR142-R150